client/my-sites/site-settings/form-writing.jsx

```diff
@@ -36,8 +36,27 @@ import ThemeEnhancements from './theme-enhancements';
 import PublishingTools from './publishing-tools';
 import QueryJetpackModules from 'components/data/query-jetpack-modules';
 import SpeedUpYourSite from './speed-up-site-settings';
+import scrollTo from 'lib/scroll-to';
 
 class SiteSettingsFormWriting extends Component {
+	componentDidMount() {
+		// Wait for page.js to update the URL, then see if we are linking
+		// directly to a section of this page.
+		setTimeout( () => {
+			if ( ! window || ! window.location ) {
+				// This code breaks everything in the tests (they hang with no
+				// error message).
+				return;
+			}
+			const hash = window.location.hash;
+			const el = hash && document.getElementById( hash.substring( 1 ) );
+			if ( hash && el ) {
+				const y = el.offsetTop - document.getElementById( 'header' ).offsetHeight - 15;
+				scrollTo( { y } );
+			}
+		} );
```

Comment:
Hi! can we talk about this? I think there are a few things we probably want to rearrange. First of all we prefer to keep `setTimeout()` of our our ~components~ codebase. It would be more preferable to fire off a Redux action here so that flow remains synchronous and so that we don't get into inconsistent states. For example, this timeout allows for us to navigate to a different page and then interrupt the scroll location. As the code shows, we already realized this pattern isn't that testable and so we've made some concessions to prevent breaking the tests.

Further we typically delegate URL handling code to the `controller.jsx` file (at worst) or some Redux middleware (at best). There's much to gain from keeping components oblivious to `window` and `document` variables and we can almost always do that if we reorganize our code.

Third, it's not clear what we are looking for in the URL. Sometimes a comment with example input can go a long way to indicating why the code is there.

```js
// https://wordpress.com/settings/my-site/#step-2
const elementId = hash && hash.substring( 1 );
const element = document.getElementById( elementId );
```

If we pass in the selected element from the outside then not only can we remove the window and document globals but we can also test this component and simulate those changes.

```js
componentDidUpdate( prevProps ) {
  if ( prevProps.selectedSelement !== this.props.selectedElement ) {
    const y = …
    scrollTo( { y } ); 
  }
}
```

Reply:
@dmsnell: thanks for the feedback - you bring up some good points. I had just re-used something that we'd implemented in the general settings screen, but would be happy to look at a better approach now that it seems like a repeating pattern. I'll address this after Tanooki finishes its support week.

Reply:
We are taking another look at this now (I wrote the original code in https://github.com/Automattic/wp-calypso/pull/25235/files#diff-bd0c17b929dd30cb6cd54259cfc10ed8R47).  The `setTimeout` was necessary because by the time this component had rendered, page.js hadn't actually updated the URL.

I am not sure if we would run into the same issue using a middleware.  Suggestions welcome while we try some things.